content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc

```diff
@@ -0,0 +1,424 @@
+    assertThat(jfrEvents).contains(event); 
+
+    ExpectedEvent = event("jdk.ThreadSleep").
+        with("time", Duration.ofSeconds(1)));
+    assertThat(jfrEvents).contains(event); 
+
+    assertThat(jfrEvents.ofType("jdk.GarbageCollection")).hasSize(1); // <5>
+  }
+}
+----
+<1> `@JfrEventTest` marks this as a JfrUnit test, activating its extension
+<2> All JFR event types to be recorded must be enabled via `@EnableEvent`
+<3> After running the test code, `awaitEvents()` must be invoked as a synchronization barrier,
+making sure all previously produced events have been received
+<4> Using the `JfrEventsAssert#event()` method, an `ExpectedEvent` instance can be created, optionally containing a set of expected attribute values, which then is asserted via `JfrEventsAssert#assertThat()`
+<5> `JfrEvents#ofType()` allows to filter on specific event types, enabling arbitrary assertions on the returned stream of ``RecordedEvent``s
+
+As we'll see in a bit, JfrUnit integrates nicely with the Java streams API,
+simplifying the filtering and aggregation of recorded events before matching them against asserted values.
+It also persists the JFR recording of each test method,
+so you can examine it after a test failure.
+
+To learn more about JfrUnit and its capabilities, take a look at the project's https://github.com/gunnarmorling/jfrunit[README].
+The project is in an early proof-of-concept stage at the moment,
+so changes to its APIs and semantics are likely.
+
+Now that you've taken the JfrUnit quick tour, let's put that knowledge into practice.
+Our example project will be the todo management Quarkus application you may already be familiar with from my earlier post about custom JFR events.
+We're going to discuss multiple examples for using JfrUnit to identify potential performance regressions.
+
+== Spotting Increased Memory Allocation
+
+At first, let's explore how to identify increased memory allocation rates.
+Typically, it's mostly library and middleware authors who are interested in this,
+because for a library such as Hibernate ORM it can make a huge difference whether a method invoked many times on a hot code path allocates a few bytes more or less memory.
+Allocating less memory means less work for the garbage collector,
+which in turn means those precious CPU cores can spend more cycles processing your actual business logic.
+
+But also for an actual application itself it can be beneficial to keep an eye -- and systematically track -- object allocations,
+as regressions there lead to increased GC pressure in turn to higher latencies and reduced through-put.
+
+The key for tracking object allocations with JFR are the `jdk.ObjectAllocationInNewTLAB` and `jdk.ObjectAllocationOutsideTLAB` events,
+which are emitted when
+
+* Object allocation triggered the creation of a new thread-local allocation buffer (TLAB) 
+* An object got allocated outside of the threads TLAB
+
+[NOTE]
+.Thread-local allocation buffers (TLAB)
```

Comment:
nice to learn about this while reading 👍

Reply:
Ha, thanks. Was also a nice learning while writing :)